prow/pubsub/subscriber/subscriper.go

```diff
@@ -0,0 +1,177 @@
+}
+
+// Subscriber handles Pub/Sub subscriptions, update metrics,
+// validates them using Prow Configuration and
+// use a kubeClientInterface to create Prow Jobs.
+type Subscriber struct {
+	ConfigAgent *config.Agent
+	Metrics     *Metrics
+	KubeClient  kubeClientInterface
+}
+
+type messageInterface interface {
+	GetAttributes() map[string]string
+	GetID() string
+	Ack()
+	Nack()
+}
+
+type pubSubMessage struct {
+	msg *pubsub.Message
+}
+
+func (m *pubSubMessage) GetAttributes() map[string]string {
+	return m.msg.Attributes
+}
+
+func (m *pubSubMessage) GetID() string {
+	return m.msg.ID
+}
+
+func (m *pubSubMessage) Ack() {
+	m.msg.Ack()
+}
+
+func (m *pubSubMessage) Nack() {
+	m.msg.Nack()
+}
+
+func extractFromAttribute(attrs map[string]string, key string) (string, error) {
+	value, ok := attrs[key]
+	if !ok {
+		return "", fmt.Errorf("unable to find %s from the attributes", key)
+	}
+	return value, nil
+}
+
+func extractPubSubMessage(attributes map[string]string) (*reporter.PubSubMessage, error) {
+	var err error
+	m := reporter.PubSubMessage{}
+	if m.Topic, err = extractFromAttribute(attributes, reporter.PubSubTopicLabel); err != nil {
+		return nil, err
+	}
+	if m.Project, err = extractFromAttribute(attributes, reporter.PubSubProjectLabel); err != nil {
+		return nil, err
+	}
+	if m.RunID, err = extractFromAttribute(attributes, reporter.PubSubRunIDLabel); err != nil {
+		return nil, err
+	}
+	return &m, nil
+}
+
+func (s *Subscriber) handleMessage(msg messageInterface, subscription string) error {
+	l := logrus.WithFields(logrus.Fields{
+		"pubsub-subscription": subscription,
+		"pubsub-id":           msg.GetID()})
+	s.Metrics.MessageCounter.With(prometheus.Labels{subscriptionLabel: subscription}).Inc()
+	l.Info("Received message")
+	eType, err := extractFromAttribute(msg.GetAttributes(), prowEventType)
+	if err != nil {
+		l.WithError(err).Error("failed to read message")
+		s.Metrics.ErrorCounter.With(prometheus.Labels{subscriptionLabel: subscription})
+		return err
+	}
+	switch eType {
+	case periodicProwJob:
+		err := s.handlePeriodicJob(l, msg, subscription)
+		if err != nil {
+			l.WithError(err).Error("failed to create Prow Periodic Job")
+			s.Metrics.ErrorCounter.With(prometheus.Labels{subscriptionLabel: subscription})
+		}
+		return err
+	}
+	err = fmt.Errorf("unsupported event type")
+	l.WithError(err).Error("failed to read message")
+	s.Metrics.ErrorCounter.With(prometheus.Labels{subscriptionLabel: subscription})
+	return err
+}
+
+func (s *Subscriber) handlePeriodicJob(l *logrus.Entry, msg messageInterface, subscription string) error {
+	l.Info("looking for periodic job")
+	name, err := extractFromAttribute(msg.GetAttributes(), prowJobName)
+	if err != nil {
+		return err
+	}
+	var periodicJob *config.Periodic
+	for _, job := range s.ConfigAgent.Config().AllPeriodics() {
+		if job.Name == name {
+			periodicJob = &job
+			break
+		}
+	}
+	if periodicJob == nil {
+		return fmt.Errorf("failed to find associated periodic job")
```

Comment:
(I would also include the name in the log)

Reply:
done